homeassistant/helpers/update_coordinator.py

```diff
@@ -0,0 +1,148 @@
+from homeassistant.util.dt import utcnow
+
+
+class UpdateFailed(Exception):
+    """Raised when an update has failed."""
+
+
+class DataUpdateCoordinator:
+    """Class to manage fetching data from single endpoint."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        name: str,
+        update_method: Callable[[], Awaitable],
+        update_interval: timedelta,
+        # How long to wait to actually do the refresh after requesting it.
+        # We wait some time so if we control multiple things, we batch requests.
+        request_refresh_delay: float,
+    ):
+        """Initialize global data updater."""
+        self.hass = hass
+        self.logger = logger
+        self.name = name
+        self.update_method = update_method
+        self.update_interval = update_interval
+        self.request_refresh_delay = request_refresh_delay
+
+        self.data: Optional[Any] = None
+
+        self._listeners: List[CALLBACK_TYPE] = []
+        self._unsub_refresh: Optional[CALLBACK_TYPE] = None
+        self._request_refresh_task: Optional[asyncio.TimerHandle] = None
+        self.failed_last_update = False
+
+    @callback
+    def async_add_listener(self, update_callback: CALLBACK_TYPE) -> None:
+        """Listen for data updates."""
+        schedule_refresh = not self._listeners
+
+        self._listeners.append(update_callback)
+
+        # This is the first listener, set up interval.
+        if schedule_refresh:
+            self._schedule_refresh()
+
+    @callback
+    def async_remove_listener(self, update_callback: CALLBACK_TYPE) -> None:
+        """Remove data update."""
+        self._listeners.remove(update_callback)
+
+        if not self._listeners and self._unsub_refresh:
+            self._unsub_refresh()
+            self._unsub_refresh = None
+
+    async def async_refresh(self) -> None:
+        """Refresh the data."""
+        if self._unsub_refresh:
+            self._unsub_refresh()
+            self._unsub_refresh = None
+
+        await self._async_do_refresh()
+        self._schedule_refresh()
+
+    @callback
+    def _schedule_refresh(self) -> None:
+        """Schedule a refresh."""
+        if self._unsub_refresh:
+            self._unsub_refresh()
+            self._unsub_refresh = None
+
+        self._unsub_refresh = async_track_point_in_utc_time(
+            self.hass, self._handle_refresh_interval, utcnow() + self.update_interval
+        )
+
+    async def _handle_refresh_interval(self, _now: datetime) -> None:
+        """Handle a refresh interval occurrence."""
+        self._unsub_refresh = None
+        await self._async_do_refresh()
+        self._schedule_refresh()
+
+    @callback
+    def async_request_refresh(self) -> None:
+        """Request a refresh.
+
+        Refresh will wait a bit to see if it can batch them.
+        """
+        if self._request_refresh_task:
+            self._request_refresh_task.cancel()
+
```

Comment:
Should we set `self._request_refresh_task` to `None` here of we have cancelled the task?

Reply:
Extracted that logic, doing that now.